### PR TITLE
[Mm-26] - fixes issue with large file sizes for watchers

### LINKED ISSUE #26: Watcher Service can easily crash because of Out of Memory while trying to read the file contents
The following line in watcher.go [watcher_crash.log](https://github.com/mattermost/mattermost-plugin-nagios/files/6770642/watcher_crash.log) tries to read the file contents in memory and then do a checksum of the contents:https://github.com/mattermost/mattermost-plugin-nagios/blob/master/internal/watcher/watcher.go#L182

This can easily cause the Watcher Service to crash because of Out of Memory reasons. 

**Steps to reproduce:**
1. Install a Nagios service on a 1GB RAM instance. 
2. Integrate Nagios plugin with Mattermost. 
3. Now try to download a 100MB file using the command in Nagios config directory:`wget https://speed.hetzner.de/100MB.bin`

Note: You can also try with bigger files from this site:https://speed.hetzner.de

4. Notice that the watcher service crashes because of the Out of Memory exception as shown in the attached logs. 

This ticket is also tracked in Jira https://mattermost.atlassian.net/browse/MM-35276

### PR DESCRIPTION
**Pull request by maisnamrajusingh**
Summary :
Fixes #26 
The Pull Request prevents the watcher from reading files of 100MB and above because the system might run out of memory. Instead of reading files, the watcher puts large files into customizable temporary folders. If the file is lower than 100 MB, the watcher reads as usual using 'ioutil'.

For large files, watcher reads using a buffer ('bufio') of 5 MB per loop. Each loop will find the difference and send the difference.

Ticket Link : #26

### CODE REVIEW

File: internal/watcher/watcher.go [hunk 1]
```diff
@@ -174,52 +182,156 @@ func (d Differential) WatchFn(path string) error {
...
 
-	contents, err := ioutil.ReadFile(path)
+	info, err := os.Stat(path)
 	if err != nil {
-		return fmt.Errorf("ioutil.ReadFile: %w", err)
+		return fmt.Errorf("error read file status %w", err)
 	}
 
-	checksum := md5.Sum(contents) //nolint:gosec
+	if info.Size() >= MaxFileSize {
```
Comment (jfrerich) [inline internal/watcher/watcher.go:192]: I think the code would benefit from creating a function that handles the case when `if info.Size() >= MaxFileSize `.  

Would it make sense to do something like the following and move the contents of the if into that separate function?

```go
if info.Size() >= MaxFileSize {
		d.WatchLargeFileFn(info, path)
		return nil
}
```

**Review by jfrerich — CHANGES_REQUESTED** [review]
hi, @maisnamrajusingh!  Thanks for the PR. 

I've highlighted a couple of suggestions that I'd like to get addressed.  I think the refactoring will help with readability and I'll ping for a second reviewer when we get this round completed.

**Comment by maisnamrajusingh** [comment]
@jfrerich  would you mind taking a look while i fix the linting issues ?

**Comment by jfrerich** [comment]
@maisnamrajusingh, thanks for the round of PR enhancements.  I'll re-review after dinner tonight :)

File: internal/watcher/watcher.go [hunk 2]
```diff
@@ -200,45 +210,19 @@ func (d Differential) WatchFn(path string) error {
...
-			contentName := "FilePath##" + TemporaryDirectory + "/" + info.Name()
-			d.previousContents[path] = []byte(contentName)
-
-			if err := d.sendDiff(path, diff); err != nil {
-				return fmt.Errorf("d.sendDiff: %w", err)
-			}
-
-			log.Printf("Sent the diff (size = %d)", len(diff))
-
+			d.WatchLargeFileFn(info, path)
```
Comment (jfrerich) [inline internal/watcher/watcher.go:221]: This function returns an error.  If error is not nil, we will need to handle it after calling the function.

File: internal/watcher/watcher.go [hunk 3]
```diff
@@ -174,52 +182,156 @@ func (d Differential) WatchFn(path string) error {
...
 		}
 
-		previousChecksum[p] = md5.Sum(b) //nolint:gosec
-		previousContents[p] = b
+		if info.Size() >= MaxFileSize {
+			log.Printf("File equal or higher than %v MB", MaxFileSize/1024/1024)
+
+			_, err = os.Stat(TemporaryDirectory)
+			if os.IsNotExist(err) {
+				err := os.MkdirAll(TemporaryDirectory, 0777)
```
[the remark this answers is not in the document]
Comment (jfrerich) [reply internal/watcher/watcher.go:302]:
```suggestion
				err := os.MkdirAll(TemporaryDirectory, FullPermissionBits)
```

**Review by jfrerich — CHANGES_REQUESTED** [review]
@hanzei I have a general question about the overall methodology of the PR.  Does copying, saving files, .etc affect High availability environments?

File: internal/watcher/watcher.go [hunk 4]
```diff
@@ -27,6 +30,27 @@ func getIgnoredExtensions(extensions []string) map[string]struct{} {
...
+var (
+	// MaxFileSize to set the maximum size of the file to be read and stored in memory,
+	// if more than that, the file will be saved to a temporary folder
+	MaxFileSize int64 = 100 * sizeMb
+
+	// MaxReadSize to set the buffer used to read the contents of the file each loop
+	MaxReadSize int64 = 5 * sizeMb
+
+	// TemporaryDirectory the location of the folder to store temporary files
+	TemporaryDirectory = os.Getenv("HOME") + "/temp/watcher"
```
Comment (hanzei) [inline internal/watcher/watcher.go:48]: Why isn't [`os.TempDir()`](https://pkg.go.dev/os#TempDir) used here?

File: internal/watcher/watcher.go [hunk 5]
```diff
@@ -169,15 +196,61 @@ func (d Differential) sendDiff(path string, diff string) error {
...
-	// bytes chunks. 1 ms should be enough.
-	time.Sleep(time.Millisecond)
+	time.Sleep(10 * time.Millisecond)
+
+	info, err := os.Stat(path)
+	if err != nil {
+		return fmt.Errorf("error read file status %w", err)
+	}
+
+	if info.Size() >= MaxFileSize {
```
Comment (hanzei) [inline internal/watcher/watcher.go:212]: Should these two code paths be unified? If there are a lot of small files, the server can still run out of memory. This would unify the use of `previousContents` which can be renamed to `previousContentPath` then.

File: internal/watcher/watcher.go [hunk 6]
```diff
@@ -231,3 +333,168 @@ func NewDifferential(
...
+	out, err := os.Create(dst)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		cerr := out.Close()
+		if err == nil {
+			err = cerr
+		}
+	}()
```
Comment (hanzei) [inline internal/watcher/watcher.go:386]: Does this correctly return the error? Isn't a named return value needed for this to work?

File: internal/watcher/watcher.go [hunk 7]
```diff
@@ -231,3 +333,168 @@ func NewDifferential(
...
+		return ""
+	}
+	return value[adjustedPos:]
+}
+
+// WatchLargeFileFn is function to read large file then puts the contents of the file into a temporary directory
+// and saves the file path to previousContents
+func (d Differential) WatchLargeFileFn(info fs.FileInfo, path string) error {
+
+	err := CopyFile(path, TemporaryDirectory+"/"+info.Name())
```
Comment (hanzei) [inline internal/watcher/watcher.go:411]: Does this work under windows?

File: internal/watcher/watcher.go [hunk 8]
```diff
@@ -231,3 +333,168 @@ func NewDifferential(
...
+	}
+	f2, err := os.Open(srcFile)
+	if err != nil {
+		return fmt.Errorf("error openning temp file: %w", err)
+	}
+	defer f1.Close()
+	defer f2.Close()
+
+	r1 := bufio.NewReader(f1)
+	r2 := bufio.NewReader(f2)
```
Comment (hanzei) [inline internal/watcher/watcher.go:456]: Why is a `bufio.Reader` needed here?

File: internal/watcher/watcher.go [hunk 9]
```diff
@@ -231,3 +333,168 @@ func NewDifferential(
...
+		if err == io.EOF {
+			eof2 = true
+		} else if err != nil {
+			return fmt.Errorf("error read chunkz watched file: %w", err)
+		}
+		content2 := buf2[:n]
+
+		if eof1 && eof2 {
+			break
+		}
```
Comment (hanzei) [inline internal/watcher/watcher.go:484]: Why does the code `break` when `EOF` is reached?

File: internal/watcher/watcher.go [hunk 10]
```diff
@@ -231,3 +333,168 @@ func NewDifferential(
...
+		content1 := buf1[:n]
+
+		// read second file
+		n, err = r2.Read(buf2)
+		if err == io.EOF {
+			eof2 = true
+		} else if err != nil {
+			return fmt.Errorf("error read chunkz watched file: %w", err)
+		}
+		content2 := buf2[:n]
```
Comment (hanzei) [inline internal/watcher/watcher.go:480]: By completely reading both files into memory a lot of memory is allocated just to be de-allocated shortly after as both files are the same in 99,999% of the cases. Is it suitable to compare the md5 of the changed files first against the md5 of the possible changed one?

File: internal/watcher/watcher_test.go [hunk 11]
```diff
@@ -149,7 +149,7 @@ func TestNewDifferential(t *testing.T) {
 			token:             "",
 		}
 
-		actual, err := NewDifferential(nil, nil, nil, "", "")
+		actual, err := NewDifferential(nil, nil, nil, "", "", "")
```
Comment (hanzei) [inline internal/watcher/watcher_test.go:152]: I would love to see a test for multiple big files.

File: internal/watcher/watcher.go [hunk 12]
```diff
@@ -231,3 +333,168 @@ func NewDifferential(
...
+		buf1 := make([]byte, MaxReadSize)
+		buf2 := make([]byte, MaxReadSize)
+		// read first file
+		n, err := r1.Read(buf1)
+		if err == io.EOF {
+			eof1 = true
+		} else if err != nil {
+			return fmt.Errorf("error read chunkz watched file: %w", err)
+		}
+		content1 := buf1[:n]
```
Comment (hanzei) [inline internal/watcher/watcher.go:471]: Why is `content` allocated every time the loop runs?

**Review by hanzei — CHANGES_REQUESTED** [review]
Given that the logic isn't straight forward I would love to see some tests for the changes.

**Comment by hanzei** [comment]
> @hanzei I have a general question about the overall methodology of the PR. Does copying, saving files, .etc affect High availability environments?

Can you please elaborate the question? How does the file watcher interact with an HA nagios environment?

**Comment by hanzei** [comment]
@jfrerich What do you think about the comment above from @amwolff ?

**Comment by jfrerich** [comment]
@srkgupta, would you mind responding to the question from @amwolff regarding the need to diff large files. This PR is a the response to a ticket found during security review and I feel it best that we get the author of the tickets response :)

**Comment by srkgupta** [comment]
Hi @amwolff 

Thanks for checking. Please find my responses inline for your comment:

> Will there ever be a need to diff such large files? The differential watcher is intended to make a diff between Nagios configuration files. Nagios configuration files are in the worst case in the range of kilobytes. Making a diff between two 5MB files or even files of arbitrary size would mean that there's something wrong (maybe we pointed the watcher at the wrong directory?). 

At the moment the watcher not only checks for specific Nagios config files, but rather all files in a specific directory. So if we want to be more selective, we can decide to monitor, i.e. notify changes for only specific list of whitelisted files and for everything else, just display a standard message like "New file found: ...."


> Plugin's server is defensive and will reject diffs larger than the message size limit:

Please note this comes into picture only after the initial diff is computed. The original issue found and reported was while computing the diff. 

> I don't think there's a point in diffs larger than the message size limit. In my opinion, the watcher should read up to, e.g. 100KB, and for larger files, it should either reject them (signalling there's something wrong in the updates channel would be perfect) or diff only that portion of the file. The first approach seems to be more explicit for the user/administrator.

I agree with the former statement, i.e to either reject/ignore when diff > 100KB and rather display a standard generic message like [here](https://github.com/mattermost/mattermost-plugin-nagios/blob/master/server/watcher.go#L23).

**Comment by maisnamrajusingh** [comment]
@jfrerich @hanzei based on the above comments, will update the PR to reflect the changes.

File: cmd/watcher/main.go [hunk 13]
```diff
@@ -12,15 +12,17 @@ import (
 )
 
 var (
-	dir   = flag.String("dir", "/usr/local/nagios/etc/", "Nagios configuration files directory")
-	url   = flag.String("url", "", "Mattermost Server address")
-	token = flag.String("token", "", "Nagios plugin token")
+	dir    = flag.String("dir", "/usr/local/nagios/etc/", "Nagios configuration files directory")
+	url    = flag.String("url", "", "Mattermost Server address")
+	token  = flag.String("token", "", "Nagios plugin token")
```
Comment (sibasankarnayak) [inline cmd/watcher/main.go:17]:
```suggestion
	token = flag.String("token", "", "Nagios plugin token")
```

File: cmd/watcher/main.go [hunk 14]
```diff
@@ -12,15 +12,17 @@ import (
 )
 
 var (
-	dir   = flag.String("dir", "/usr/local/nagios/etc/", "Nagios configuration files directory")
-	url   = flag.String("url", "", "Mattermost Server address")
-	token = flag.String("token", "", "Nagios plugin token")
+	dir    = flag.String("dir", "/usr/local/nagios/etc/", "Nagios configuration files directory")
+	url    = flag.String("url", "", "Mattermost Server address")
```
Comment (sibasankarnayak) [inline cmd/watcher/main.go:16]:
```suggestion
	url   = flag.String("url", "", "Mattermost Server address")
```

File: cmd/watcher/main.go [hunk 15]
```diff
@@ -12,15 +12,17 @@ import (
 )
 
 var (
-	dir   = flag.String("dir", "/usr/local/nagios/etc/", "Nagios configuration files directory")
-	url   = flag.String("url", "", "Mattermost Server address")
-	token = flag.String("token", "", "Nagios plugin token")
+	dir    = flag.String("dir", "/usr/local/nagios/etc/", "Nagios configuration files directory")
```
Comment (sibasankarnayak) [inline cmd/watcher/main.go:15]:
```suggestion
	dir   = flag.String("dir", "/usr/local/nagios/etc/", "Nagios configuration files directory")
```

**Review by sibasankarnayak — CHANGES_REQUESTED** [review]

File: internal/watcher/watcher.go (hunk 10 above)
Comment (sibasankarnayak) [reply internal/watcher/watcher.go:479]: @hanzei is this what you are referring to https://codereview.stackexchange.com/questions/109942/comparing-md5-sum-to-text-from-file

Comment (hanzei) [reply internal/watcher/watcher.go:479]: I don't quite get your question. The link you posted shows an example implementation of a similar issue, but this one here is simpler. Both buffers are bytes slices.

**Comment by hanzei** [comment]
Discussed offline: Tests are a must have this feature contains some complexity and it's easy to write tests.

File: internal/watcher/watcher.go (hunk 10 above)
Comment (maisnamrajusingh) [reply internal/watcher/watcher.go:479]: @hanzei  I can only think of two days to check the difference 
1. we compare the files the first time and store the checksum inside a store and then compare it or update it depending on the use case in the logic. 
2. the implementation here where we compare the files. 

Do you have any other ways that you think would be more performant ? or shall we go for 1 above ?

File: internal/watcher/watcher.go (hunk 7 above)
Comment (hanzei) [reply internal/watcher/watcher.go:410]: It could run on Windows, right?

**Comment by lufty-demansol** [comment]
@jfrerich @hanzei @maisnamrajusingh I have added changes to only allow `.cfg` files to be scanned by differential watcher. I read in Nagios documentation that the extension for configuration files is `.cfg`.

File: internal/watcher/watcher.go (hunk 7 above)
Comment (maisnamrajusingh) [reply internal/watcher/watcher.go]: updated this earlier to use the filepath method

File: internal/watcher/watcher.go (hunk 8 above)
Comment (lufty-demansol) [reply internal/watcher/watcher.go]: @hanzei this code has been removed

File: internal/watcher/watcher.go (hunk 9 above)
Comment (lufty-demansol) [reply internal/watcher/watcher.go]: @hanzei this code has been removed

File: internal/watcher/watcher.go (hunk 10 above)
Comment (lufty-demansol) [reply internal/watcher/watcher.go]: @hanzei @maisnamrajusingh this code has been removed

File: internal/watcher/watcher_test.go (hunk 11 above)
Comment (lufty-demansol) [reply internal/watcher/watcher_test.go]: @hanzei this changes has been reverted by https://github.com/mattermost/mattermost-plugin-nagios/pull/31/commits/9a459726ef24c32fec41c57f995b74f477cf03e7

**Comment by maisnamrajusingh** [comment]
@hanzei the changes here are different from what was initially proposed and this does look a bit better. Would it make sense to send this out as a seperate PR ?

File: internal/watcher/watcher.go [hunk 16]
```diff
@@ -169,15 +161,16 @@ func (d Differential) sendDiff(path string, diff string) error {
...
-	if _, ok := d.ignoredExtensions[filepath.Ext(path)]; ok {
+	if _, ok := d.allowedExtensions[filepath.Ext(path)]; !ok {
 		return nil
 	}
 
 	// This is to allow for changes to propagate on the filesystem. If the file
 	// is large, the write won't be atomic. It will happen in, for example, 4096
 	// bytes chunks. 1 ms should be enough.
-	time.Sleep(time.Millisecond)
+	time.Sleep(10 * time.Millisecond)
```
Comment (hanzei) [inline internal/watcher/watcher.go:173]: Why was this changed?

File: internal/watcher/watcher.go [hunk 17]
```diff
@@ -206,9 +199,14 @@ func (d Differential) WatchFn(path string) error {
 
 // NewDifferential returns initialized Differential.
 func NewDifferential(
-	ignoredExtensions, initialFilePaths []string,
+	allowedExtensions map[string]bool,
```
Comment (hanzei) [inline internal/watcher/watcher.go:202]: 1/5 that I liked the old API as the `map` is an implementation detail and a string slice is enough.

File: internal/watcher/watcher.go [hunk 18]
```diff
@@ -114,18 +104,20 @@ func WatchDirectories(
 // Differential implements WatchFuncProvider. Use NewDifferential to initialize
 // Differential.
 type Differential struct {
-	ignoredExtensions map[string]struct{}
+	allowedExtensions map[string]bool
```
Comment (hanzei) [inline internal/watcher/watcher.go:107]: It's idiomatic to use a `map[string]struct{}` if the map is only used for lookup
```suggestion
	allowedExtensions map[string]struct{}
```

File: internal/watcher/watcher.go [hunk 19]
```diff
@@ -63,6 +52,7 @@ func GetAllInDirectory(dir string, ignoredExtensions []string) (
 	return files, directories, nil
 }
 
+// WatchFuncProvider for interfacing WatchFn for WatchDirectories
```
Comment (hanzei) [inline internal/watcher/watcher.go:55]: What does this sentence mean?

**Review by hanzei — CHANGES_REQUESTED** [review]
Thanks for the update :+1: 

Two blocker from my side:
1. As mentioned in https://github.com/mattermost/mattermost-plugin-nagios/pull/31#issuecomment-917878613 the code should reject files how are larger the 100KB.
2. Could you please add a test 1. and the ignore case?

**Comment by hanzei** [comment]
@maisnamrajusingh The PR has the whole conversation history. Let's preserve that and keep using it.

File: internal/watcher/watcher.go (hunk 16 above)
Comment (lufty-demansol) [reply internal/watcher/watcher.go:173]: @hanzei the original code in `master` branch uses 1 ms. 10 ms is from PR's previous changes.

**Comment by lufty-demansol** [comment]
> 1. #31 (comment)

@hanzei as mentioned by @amwolff [here](https://github.com/mattermost/mattermost-plugin-nagios/pull/31#pullrequestreview-738819560), the differential watcher is intended to make a diff between Nagios configuration files so I think filtering by extension is enough and no need to check file size.

File: internal/watcher/watcher.go (hunk 16 above)
Comment (lufty-demansol) [reply internal/watcher/watcher.go]: Reverted by https://github.com/mattermost/mattermost-plugin-nagios/pull/31/commits/6761c161991e6304f20e77fadd6712ba758b28b7

File: internal/watcher/watcher.go (hunk 17 above)
Comment (lufty-demansol) [reply internal/watcher/watcher.go]: Map usage changed to string slice: https://github.com/mattermost/mattermost-plugin-nagios/pull/31/commits/ef27e99c2f5e1082e5c5de658675f9d2d2f0e4ee

File: internal/watcher/watcher.go (hunk 18 above)
Comment (lufty-demansol) [reply internal/watcher/watcher.go]: Map usage changed to string slice: https://github.com/mattermost/mattermost-plugin-nagios/pull/31/commits/ef27e99c2f5e1082e5c5de658675f9d2d2f0e4ee

File: internal/watcher/watcher.go (hunk 19 above)
Comment (lufty-demansol) [reply internal/watcher/watcher.go]: Fixed by https://github.com/mattermost/mattermost-plugin-nagios/pull/31/commits/e9d81e1dc1d8a7df42365816b2c1900c497136bd

Comment (lufty-demansol) [reply internal/watcher/watcher.go]: > Thanks for the update 👍
> 
> Two blocker from my side:
> 
> 1. As mentioned in [#31 (comment)](https://github.com/mattermost/mattermost-plugin-nagios/pull/31#issuecomment-917878613) the code should reject files how are larger the 100KB.
> 2. Could you please add a test 1. and the ignore case?

Added ignore test: https://github.com/mattermost/mattermost-plugin-nagios/pull/31/commits/ec950370c5676edb753bb14d7d985d1cc6b18c02

**Comment by lufty-demansol** [comment]
> Great work on the tests 💯
> 
> 2/5 that I would still like to see a safety check for the file size. If an admin accidentally uploads a very large file, the watcher should not crash.

Added by https://github.com/mattermost/mattermost-plugin-nagios/pull/31/commits/3850518e8c29dc132b3aadf80e11b17bb0743a65

File: internal/watcher/watcher.go [hunk 20]
```diff
@@ -114,63 +164,30 @@ func WatchDirectories(
...
-	ignoredExtensions map[string]struct{}
+	allowedExtensions []string
 	previousChecksum  map[string][16]byte
 	previousContents  map[string][]byte
 	client            *http.Client
 	url, token        string
+	diffSender        DiffSender
 }
 
+// Change struct for send Difference
```
[the remark this answers is not in the document]
Comment (lufty-demansol) [reply internal/watcher/watcher.go]: Fixed by https://github.com/mattermost/mattermost-plugin-nagios/pull/31/commits/7926fb3d106f87e7cadfa484abd16065e92faa0f

File: internal/watcher/watcher_test.go [hunk 21]
```diff
@@ -192,24 +204,89 @@ func TestNewDifferential(t *testing.T) {
...
+	}
+
+	assert.Equal(t, 0, len(d.previousChecksum))
+}
+
+type mockDiffSender struct{}
+
+func (d mockDiffSender) Send(path string, diff string) error {
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (lufty-demansol) [reply internal/watcher/watcher_test.go]: Fixed by https://github.com/mattermost/mattermost-plugin-nagios/pull/31/commits/c0b296f27e062f258e5e1cdec426cb81dd066bcc

File: internal/watcher/watcher.go [hunk 22]
```diff
@@ -223,11 +245,16 @@ func NewDifferential(
...
 
 	return Differential{
-		ignoredExtensions: getIgnoredExtensions(ignoredExtensions),
+		allowedExtensions: allowedExtensions,
 		previousChecksum:  previousChecksum,
 		previousContents:  previousContents,
 		client:            httpClient,
 		url:               url,
 		token:             token,
+		diffSender: RemoteDiffSender{
```
[the remark this answers is not in the document]
Comment (lufty-demansol) [reply internal/watcher/watcher.go]: @hanzei the purpose to use interface is to allow diff sender to be tested/mocked.

**Review by hanzei — APPROVED** [review]
Awesome :tada:

**Review by sibasankarnayak — APPROVED** [review]
LGTM!

**Comment by hanzei** [comment]
@jfrerich @amwolff Please take another look when you have the time.

**Review by jfrerich — APPROVED** [review]
LGTM!  Thanks, @maisnamrajusingh!

**Comment by hanzei** [comment]
@dipak-demansol Do you want to test this PR on it's own or as part of the release testing?

**Comment by dipak-demansol** [comment]
> @dipak-demansol Do you want to test this PR on it's own or as part of the release testing?

@hanzei its bets to test it in this PR because i don't want that at the time of release we report a issue and then developer will fix, but Currently we're still working on release testing so this week is not possible to test this PR also i need Dylan help on this to test.

**Comment by hanzei** [comment]
@dipak-demansol gentle reminder to review this PR

**Comment by hanzei** [comment]
@dipak-demansol What is the status of testing this PR?

**Comment by dipak-demansol** [comment]
> @dipak-demansol What is the status of testing this PR?

@hanzei sorry to say but there is no progress on that by me, last week i was working on some other PR plus did the release testing of TODO plugin, in this week i'll try my best to complete the testing this PR.

**Comment by DHaussermann** [comment]
@maisnamrajusingh We have setup for Nagios but it's unclear how exactly this pluign should function.
I'm curious if there is some type of connect or authentication mechanism. Have you been able to setup the plugin and get it to return data?

**Comment by hanzei** [comment]
According to the manifest (https://github.com/mattermost/mattermost-plugin-nagios/blob/f1449342602ed1db6087179a6baf0c420e2b4d56/plugin.json#L22-L33) you simply need to set a `NagiosURL`. Nothing more seems to be required.

**Comment by DHaussermann** [comment]
@maisnamrajusingh or @hanzei 

Here is the current status on Nagios testing.
- Dipak has an instance of Nagio up and running that has report and data on hosts
- We are trying to configure and use the plugin on a Mattermost server with this instance of Nagios
- When trying to use the plugin to set up alerts the response is always `Getting log unsuccessful`

I believe what's happening is there has been no authentication configured in his setup. There is a token available in the Mattermost cofig but, nowhere on the Nagios side to accept it. All mentions of a token in the documentation are in the context of watcher. 

Is it possible the plugin only communicates with watcher and not Nagios directly? Regardless are we sure watcher is optional for setup and the documentation states?

**Comment by dipak-demansol** [comment]
@maisnamrajusingh @hanzei nagios and watcher is up and running still im getting this kind of error. 

image

below is my watcher file as per the github doc
image

**Comment by dipak-demansol** [comment]
it was hard to understand the github doc, setup the Configure the watcher section, pls improve it. 
image

File: internal/watcher/watcher_test.go [hunk 23]
```diff
@@ -192,24 +204,131 @@ func TestNewDifferential(t *testing.T) {
...
+
+func TestDifferentialLargeFile(t *testing.T) {
+	baseDir, err := ioutil.TempDir("", "watcher_test_*")
+	if err != nil {
+		t.Fatalf("ioutil.TempDir: %v", err)
+	}
+
+	defer os.RemoveAll(baseDir)
+
+	testContent, err := ioutil.ReadFile("testdata/test-large-file.cfg")
```
Comment (amwolff) [inline internal/watcher/watcher_test.go:312]: nit: the large file could be generated on demand instead of including it in the repository :)

File: internal/watcher/watcher.go [hunk 24]
```diff
@@ -179,6 +196,15 @@ func (d Differential) WatchFn(path string) error {
...
 
+	fileInfo, err := os.Stat(path)
+	if err != nil {
+		return fmt.Errorf("os.Stat: %w", err)
+	}
+
+	if fileInfo.Size() > 100*1024 {
+		return nil
+	}
+
```
Comment (amwolff) [inline internal/watcher/watcher.go:207]: nit: it might be nice to log we rejected the file based on these criteria

**Review by amwolff — APPROVED** [review]
Nice work! I really like new tests

**Comment by hanzei** [comment]
@DHaussermann @dipak-demansol Could you please go into more details on when you are seeing the above posted error? Can you access nagios (via the web)? Do any of the slash commands produce a non-error?

Secondly: What nagios server is used? Can I get access to it?

**Comment by hanzei** [comment]
Re https://github.com/mattermost/mattermost-plugin-nagios/pull/31#issuecomment-1141107431: Good suggestion, fixed that in [`7233e2e` (#40)](https://github.com/mattermost/mattermost-plugin-nagios/pull/40/commits/7233e2e49624375d71293ec9b1e282f2425c8959)

**Comment by hanzei** [comment]
Closing in favor of https://github.com/mattermost/mattermost-plugin-nagios/pull/40
